File: combn.cpp
#include <stdio.h>
#include <stdlib.h>
#include <iostream>
#include <vector>

using namespace std;

vector<int> values;
vector<int> combination;
vector<vector <int> > allCombinations;
int *allCombs;
int *numEntriesPerLevel;

double choose(int n, int k) {
    if(k == 0) return 1; 
    return (n * choose(n - 1, k - 1)) / k;
}

void print1d(const vector<int>& v) {
	static int count = 0;
	for(int i = 0; i < v.size(); i++) {
		cout << v[i] << " ";
	}
	cout << endl;
}

void print2d(const vector<vector<int> > &v) {
	for(int i = 0; i < v.size(); i++) {
		for(int j = 0; j < v[i].size(); j++) {
			cout << v[i][j] << " ";
		}
		cout << endl;
	}
}

void addComb(vector<vector<int> > &v, vector<int> &v2) {
	v.push_back(v2);
}

void findCombs(int offset, int k) {
	if (k == 0) {
		// print1d(combination);
		addComb(allCombinations, combination);
		return;
	}
	for(int i = offset; i <= values.size() - k; ++i) {
		combination.push_back(values[i]);
		findCombs(i + 1, k - 1);
		combination.pop_back();
	}
}

void findEntriesPerLevel(int *array, int x, int m) {
	for(int i = 0; i < x - m + 1 ; i++) {
		array[i] = choose(x - i - 1, m - 1);
	}
}

int * combn(int x, int m) {
	int num = choose(x,m);
	for(int i = 0; i < x; ++i)
		values.push_back(i+1);

	int size = choose(x,m); 
	allCombs = new int[size * m];
	numEntriesPerLevel = new int[x - m + 1];
	findEntriesPerLevel(numEntriesPerLevel, x, m);

	for(int i = 0; i < x - m + 1; i++){
		cout << numEntriesPerLevel[i] << " ";	}

	#pragma omp parallel for
	for(int i = 1; i <= x - m + 1; i++){	
		combination.push_back(i);
		findCombs(i, m - 1);
		combination.pop_back();
	}
	#pragma omp barrier
	// print2d(allCombinations);

	int *ret = new int[num * m];
	return ret;
}



int main (int argc, char** argv) {

	int x = 5;
	int m = 3;
	int * vals;
    vals = combn(x,m);
    int count = 0;

    return 0;
}

File: combn.R
#Arjun Bharadwaj abharadwaj@ucdavis.edu
#Bijan Agahi bsagahi@ucdavis.edu
#Stefan Peterson stpeterson@ucdavis.edu

#Import the existing libraries(mainly snow nad parallel)
library("snow")

combn_R <- function(x, m, FUN = NULL, simplify = TRUE) {
  # Determine the total size
  size <- choose(x, m)
  # Generate from 1 to n range
  values <- seq_len(x)
  # All combinations
  allCombs <- vector(mode = "numeric", length = size * m)
  # Number of entries Per Level
  numEntriesPerLevel <- vector(mode = "numeric", length = x - m + 1)
  # All permutations
  allCombinations <- list()
  # Setup the initial stuff
  findEntriesPerLevel <- function() {
    # Indices in R start at 1
    numEntriesPerLevel[1] <- 0
    for(i in 2:(x - m + 1)) {
      numEntriesPerLevel[i] <- numEntriesPerLevel[i - 1] + (choose(x - i + 1, m - 1) * m)
    }
    numEntriesPerLevel
  }

  push_back <- function(combination, value) {
    combination[length(combination)+1] <- value
  }

  pop_back <- function(combination) {
    combination <- combination[-(length(combination))]
  }

  position <- 1

  findCombs <- function(offset, k, v, combination) {
    if(k == 0) {
      allCombinations[[position]] <<- combination
      position <<- position + 1

      return(0)
    }
    for(i in offset:((length(values) - k)+1) ) {
      combination <- c(combination, values[i])
      findCombs(i + 1, k - 1, v, combination)
      combination <- pop_back(combination)
    }
  }

  numEntriesPerLevel <- findEntriesPerLevel()
  #print(numEntriesPerLevel)

  # This is the main loop
  for(i in 1:(x - m + 1)) {
    # List of vectors. In C++, 2d Vector
    levelCombinations <- list()
    # Our local 1d array
    combination <- c()
    #i <- 1
    combination <- c(combination, i)
    # Call to recursive function
    findCombs(i + 1, m - 1, levelCombinations, combination)
    #print(combination)
    combination <- pop_back(combination)
  }
  allCombinations
}

